File: crates/ecaa-conformance/tests/fixtures/minimal-package-real/runtime/plotting_r/stages/spatial_clustering.R
# R-side spatial_clustering stage: tissue_overlay, morans_i_scatter,
# neighborhood_enrichment. Mirrors lib/plotting/stages/spatial_clustering.py.
#
# Plan reference: §S13.6 — spatial-transcriptomics figures. Tissue
# image underlay is best-effort (PNG/JPG read via the `png`/`jpeg`
# packages when available); when absent we fall through to the
# image-less overlay shape, matching the Python soft-fallback.

if (!exists("swfc_register_figure")) {
  stop("source runtime/plotting_r/core.R before this stage module")
}

.spatial_load_image <- function(ctx) {
  p <- .swfc_manifest_path(ctx$manifest, ctx$outputs_dir, "image")
  if (is.null(p)) return(NULL)
  ext <- tolower(tools::file_ext(p))
  img <- tryCatch({
    if (ext == "png" && requireNamespace("png", quietly = TRUE)) {
      png::readPNG(p)
    } else if (ext %in% c("jpg", "jpeg") && requireNamespace("jpeg", quietly = TRUE)) {
      jpeg::readJPEG(p)
    } else {
      NULL
    }
  }, error = function(e) NULL)
  img
}

swfc_register_figure("spatial_clustering", "tissue_overlay", function(ctx) {
  p <- .swfc_manifest_path(ctx$manifest, ctx$outputs_dir, "coords_table")
  if (is.null(p)) stop("manifest.coords_table required")
  df <- .swfc_load_tsv(p, list(
    x = c("x", "x_pixel", "imagecol"),
    y = c("y", "y_pixel", "imagerow"),
    value = c("value", "cluster", "domain", "expression")
  ))
  if (is.null(df)) stop(sprintf("unparseable coords table: %s", p))
  image <- .spatial_load_image(ctx)
  swfc_tissue_overlay_r(df, image = image, title = "Tissue overlay")
})

swfc_register_figure("spatial_clustering", "morans_i_scatter", function(ctx) {
  p <- .swfc_manifest_path(ctx$manifest, ctx$outputs_dir, "morans_i_table")
  if (is.null(p)) stop("manifest.morans_i_table required")
  df <- .swfc_load_tsv(p, list(
    gene = c("gene", "feature", "symbol"),
    morans_i = c("morans_i", "i", "morans"),
    p_value = c("p_value", "pvalue", "p", "morans_p")
  ))
  if (is.null(df)) stop(sprintf("unparseable morans_i table: %s", p))
  swfc_morans_i_scatter_r(df, title = "Moran's I")
})

swfc_register_figure("spatial_clustering", "neighborhood_enrichment",
                     function(ctx) {
  p <- .swfc_manifest_path(ctx$manifest, ctx$outputs_dir, "neighborhood_table")
  if (is.null(p)) stop("manifest.neighborhood_table required")
  df <- .swfc_load_tsv(p, list(
    source = c("source", "source_type", "from"),
    target = c("target", "target_type", "to"),
    score = c("score", "z", "enrichment")
  ))
  if (is.null(df)) stop(sprintf("unparseable neighborhood table: %s", p))
  swfc_neighborhood_enrichment_r(df, title = "Neighborhood enrichment")
})

File: crates/ecaa-conformance/tests/fixtures/minimal-package-real/runtime/plotting/stages/spatial_clustering.py
"""Spatial clustering stage — spatial-transcriptomics taxonomy.
Renders the per-spot tissue overlay, the Moran's I scatter for
spatially-variable genes, and the neighborhood-enrichment heatmap.

Plan reference: §S13 (spatial-transcriptomics taxonomy stage
`spatial_clustering`).

Manifest contract:
- ``coords_table``: TSV with `x`, `y`, `value`. Value is the
  per-spot statistic (cluster id, gene expression, spatial-domain
  membership) to encode by colormap.
- ``image``: optional path to a tissue image (PNG/JPG) to draw as
  the underlay. When absent, falls back to a clean scatter without
  background.
- ``morans_i_table``: TSV with `gene`, `morans_i`, `p_value`.
- ``neighborhood_table``: TSV with `source`, `target`, `score`.
"""

from __future__ import annotations

from pathlib import Path
from typing import Optional

import numpy as np

from ..core import (
    FigureContext,
    morans_i_scatter,
    neighborhood_enrichment,
    register_figure,
    stage_registry,
    tissue_overlay,
)
from ._shared import load_tsv_columns, manifest_path

FIGURES = stage_registry("spatial_clustering")


def _load_image(ctx: FigureContext) -> Optional[np.ndarray]:
    p = manifest_path(ctx.manifest, ctx.outputs_dir, "image")
    if p is None:
        return None
    try:
        from matplotlib.image import imread

        return imread(p)
    except (ImportError, OSError):
        return None


@register_figure(FIGURES, "tissue_overlay")
def tissue_overlay_fig(ctx: FigureContext, out: Path) -> Optional[Path]:
    p = manifest_path(ctx.manifest, ctx.outputs_dir, "coords_table")
    if p is None:
        raise FileNotFoundError("manifest.coords_table required")
    cols = load_tsv_columns(
        p,
        {
            "x": ("x", "x_pixel", "imagecol"),
            "y": ("y", "y_pixel", "imagerow"),
            "value": ("value", "cluster", "domain", "expression"),
        },
    )
    if cols is None:
        raise FileNotFoundError(f"unparseable coords table: {p}")
    image = _load_image(ctx)
    return tissue_overlay(
        coords_df=cols,
        image=image,
        title="Tissue overlay",
        out=out,
    )


@register_figure(FIGURES, "morans_i_scatter")
def morans_i_scatter_fig(ctx: FigureContext, out: Path) -> Optional[Path]:
    p = manifest_path(ctx.manifest, ctx.outputs_dir, "morans_i_table")
    if p is None:
        raise FileNotFoundError("manifest.morans_i_table required")
    cols = load_tsv_columns(
        p,
        {
            "gene": ("gene", "feature", "symbol"),
            "morans_i": ("morans_i", "I", "morans"),
            "p_value": ("p_value", "pvalue", "p", "morans_p"),
        },
    )
    if cols is None:
        raise FileNotFoundError(f"unparseable morans_i table: {p}")
    return morans_i_scatter(
        frame=cols, title="Moran's I", out=out
    )


@register_figure(FIGURES, "neighborhood_enrichment")
def neighborhood_enrichment_fig(ctx: FigureContext, out: Path) -> Optional[Path]:
    p = manifest_path(ctx.manifest, ctx.outputs_dir, "neighborhood_table")
    if p is None:
        raise FileNotFoundError("manifest.neighborhood_table required")
    cols = load_tsv_columns(
        p,
        {
            "source": ("source", "source_type", "from"),
            "target": ("target", "target_type", "to"),
            "score": ("score", "z", "enrichment"),
        },
    )
    if cols is None:
        raise FileNotFoundError(f"unparseable neighborhood table: {p}")
    return neighborhood_enrichment(
        frame=cols, title="Neighborhood enrichment", out=out
    )
